sdks/java/core/src/main/java/org/apache/beam/sdk/io/DefaultFilenamePolicy.java

```diff
@@ -72,52 +74,106 @@
+      String defaultTemplate =
+          windowedWrites ? DEFAULT_WINDOWED_SHARD_TEMPLATE : DEFAULT_UNWINDOWED_SHARD_TEMPLATE;
+      return new Params(baseFilename,
+          firstNonNull(shardTemplate, defaultTemplate),
+          firstNonNull(filenameSuffix, ""));
+    }
   }
 
   /**
-   * A helper function to construct a {@link DefaultFilenamePolicy} using the standard filename
-   * parameters, namely a provided {@link ResourceId} for the output prefix, and possibly-null
-   * shard name template and suffix.
-   *
-   * <p>Any filename component of the provided resource will be used as the filename prefix.
-   *
-   * <p>If provided, the shard name template will be used; otherwise
-   * {@link #DEFAULT_UNWINDOWED_SHARD_TEMPLATE} will be used for non-windowed file names and
-   * {@link #DEFAULT_WINDOWED_SHARD_TEMPLATE} will be used for windowed file names.
+   */
+  public static class ParamsCoder extends AtomicCoder<Params> {
+    private static final ParamsCoder INSTANCE = new ParamsCoder();
+    private Coder<String> stringCoder = StringUtf8Coder.of();
+
+    public static ParamsCoder of() {
+      return INSTANCE;
+    }
+
+    @Override
+    public void encode(Params value, OutputStream outStream)
+        throws IOException {
+      if (value == null) {
+        throw new CoderException("cannot encode a null value");
+      }
+      stringCoder.encode(value.baseFilename.get().toString(), outStream);
+      stringCoder.encode(value.shardTemplate, outStream);
+      stringCoder.encode(value.suffix, outStream);
+    }
+
+    @Override
+    public Params decode(InputStream inStream) throws IOException {
+      ResourceId prefix = FileBasedSink.convertToFileResourceIfPossible(
+          stringCoder.decode(inStream));
+      String shardTemplate = stringCoder.decode(inStream);
+      String suffix = stringCoder.decode(inStream);
+      return new Params(prefix, shardTemplate, suffix);
+    }
+  }
+
+  /**
+   * Constructs a new {@link DefaultFilenamePolicy}.
    *
-   * <p>If provided, the suffix will be used; otherwise the files will have an empty suffix.
+   * @see DefaultFilenamePolicy for more information on the arguments to this function.
    */
-  public static DefaultFilenamePolicy constructUsingStandardParameters(
-      ValueProvider<ResourceId> outputPrefix,
-      @Nullable String shardTemplate,
-      @Nullable String filenameSuffix,
-      boolean windowedWrites) {
-    // Pick the appropriate default policy based on whether windowed writes are being performed.
-    String defaultTemplate =
-        windowedWrites ? DEFAULT_WINDOWED_SHARD_TEMPLATE : DEFAULT_UNWINDOWED_SHARD_TEMPLATE;
-    return new DefaultFilenamePolicy(
-        NestedValueProvider.of(outputPrefix, new ExtractFilename()),
-        firstNonNull(shardTemplate, defaultTemplate),
-        firstNonNull(filenameSuffix, ""));
+  @VisibleForTesting
+  DefaultFilenamePolicy(Params params) {
+    this.baseFilename = params.baseFilename;
```

Comment:
Changed to store instance of params.

the purpose of Params is so that dynamic policies can use a Params as DestinationT. Are you suggesting that the DefaultFilenamePolicy itself be the DestinationT instead?